### ISSUE TITLE
Add RandomState support for the sklearn wrapper

### ISSUE THREAD

**Issue by ekerazha** [Microsoft/LightGBM#730]
Some libs that expect a sklearn-compatible estimator, pass a random_state parameter.

The random_state parameter is similar to the seed parameter of LightGBM, but it can also receive a RandomState instance.

    random_state : int, RandomState instance or None; default=None
        If int, random_state is the seed used by the random number generator;
        If RandomState instance, random_state is the random number generator;
        If None, the random number generator is the RandomState instance used
        by `np.random`.

At the moment, if I pass a random_state parameter with a RandomState instance, I get a `Unknown type of parameter` exception from **line 134** in **basic.py**

Would it be possible to add support for a sklearn-compatible random_state parameter for the sklearn wrapper?

**Comment by guolinke** [comment]
@wxchan

**Comment by wxchan** [comment]
@ekerazha all our current random numbers are generated in c++ code, so do you want to use RandomState to generate random number, or only the random seed matters (the seed will pass to c++ and still generate random number in c++)?

**Comment by ekerazha** [comment]
@wxchan I think it should be ok to just take the seed, however it is not possible to get the original seed from a RandomState instance (see: https://stackoverflow.com/questions/32172054/how-can-i-retrieve-the-current-seed-of-numpys-random-number-generator ). Maybe we can still get a value (although it's not the original seed) calling get_state() (see: https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.get_state.html ) and pass this value to the c++ code.

So the approach could be this:
1) If random_state is a int value, pass it as seed to the c++ code (just like for the seed parameter).
2) If random_state is a RandomState instance, call get_state() and derive from the returned value (how?) a seed for the c++ code. Ideally we would directly use RandomState to generate random numbers, but I don't know if this is feasible.

**Comment by wxchan** [comment]
@ekerazha The simplest way is to generate a random number as the seed, but I think it won't be what you want. (because it makes no differences as passing an integer.) But actually for current codes, even if you pass an integer, it is still used to generate 4 different random number for different purposes (bagging, sub-feature etc).

**Comment by ekerazha** [comment]
@wxchan I think I'd go for the two steps I listed above, so:

- If random_state is a int value, just pass that value as seed
- If random_state is a RandomState instance, call get_state() and derive a int value from that (crc32 on some value returned by get_state()?) to use as seed

What do you think about this?

**Comment by wxchan** [comment]
@ekerazha what is the different between `crc32 on RandomState.get_state() (which returns tuple(str, ndarray of 624 uints, int, int, float))` and `a random number generated by RandomState, like RandomState.randint(MAX_INT)`? Or can you describe where you want to use RandomState instead of an integer?

**Comment by ekerazha** [comment]
@wxchan well, if for the same RandomState original seed, randint(MAX_INT) always returns the same value, I think it should be ok to use it... maybe the crc32 on internal state approach is an overcomplication :-)

**Comment by wxchan** [comment]
@ekerazha sorry I am a little distracted by other things. Your solution is right and better. But the actual key question here is, why don't you just use an integer, what is the necessity to use RandomState. You can simply pass an integer to seed, using RandomState doesn't make thing easier in my opinion.

**Comment by ekerazha** [comment]
@wxchan Because some libs that rely on sklearn estimators, expect a sklearn-compatible behavior for the random_state parameter.

**Comment by wxchan** [comment]
But they can't accept integer?

**Comment by wxchan** [comment]
I think this feature is acceptable, but the solution is not elegant enough (not like we can get seed directly from RandomState). And it's not something we must add. What's the solution of those sklearn-compatible estimators?
